### DISCUSSION TITLE
Defining system tables / schema for custom extensions

### DISCUSSION THREAD

**Discussion by AustinPhillipTaylor** [directus/directus#8350] [Feature Requests]
Branching off from #3611

There are plenty of cases where a module extension (feasibly even an interface or panel) will require its own tables in the database. Currently, the best method of achieving this is simply creating a hidden collection, where the fields are set to those required by the extension. This is great because it allows extension developers to continue to access their data through the REST / GraphQL APIs provided by the collections system. 

However, this introduces a few issues. 

Foremost, if the extension is strictly reliant on the defined schema, it's not desirable that a privileged user can go into the Data Models section of the UI and accidentally change a field, without realizing that they'd be breaking functionality. 

In most cases the system should work as originally intended by Directus, as a mirror of the database. I do believe adhering to that directive creates the least friction between the frontend and database. However, I foresee cases where the structure / schema of one table will be reliant on a second table. Much like how the columns in a collection table are tightly coupled with the field definitions in the `directus_fields` table. 

Notwithstanding everything above, I also believe that utilizing only collections in extensions can lead to a lot mess in the Data Model section of the app. I haven't thought through everything just yet, but in the module I'm currently working on, I'd require at least 5 different collections to achieve my desired functionality. With just one module, this isn't a terrible issue. But as the ecosystem grows, especially with the eventual release of the Extension Marketplace, if every module were to utilize collections for data storage, and they all displayed them in the Data Model section, it'd get rather unruly. 

With all of that said, I propose a two-fold solution. 

**1)** Maintaining the functionality of the collections system is a must for extensions, but it'd be useful to add a column to `directus_collections`, something like `managed`, that would accept either `false`, to display as a normal collection, or a string corresponding to an extension's ID, where the collection schema would be hidden from the Data Model UI. This could just as simply be a true or false, but I imagine having the extension ID available could be useful down the line. For instance, if the extension is deleted or disabled, we could unhide the collection(s) from the Data Model UI, or even prompt the user if they want to delete the tables. 

**2)** Furthermore, we should implement an API to programmatically create/read/update extension-specific tables. The tables could each be prefixed by the extension ID, or a prefix otherwise defined in the extension definition object. These tables could then be automatically provided/injected to the extension, so they can easily be queried. 

I'd love to hear any thoughts, or further ideas on the matter.

**Comment by rijkvanzanten** [comment]
@nickrum Any thoughts on point 2? We might be able to rely on a very similar structure we got going on with the schema migration file in order to define a schema for custom extensions 🤔 We could then include that into the schema apply command as well to auto-migrate it in

**Reply by nickrum** [reply]
I like the idea. It might be tricky to merge the snapshots, though. Alternatively, we could rely on custom migrations which should already work.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
